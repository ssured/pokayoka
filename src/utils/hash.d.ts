export default function hash(input: string): string;
