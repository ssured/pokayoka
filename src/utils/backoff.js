class Backoff {
  constructor(noDelay = 0, minDelay = 100, maxDelay = 2000, factor = 2) {
    this.noDelay = this.current = noDelay;
    this.minDelay = minDelay;
    this.maxDelay = maxDelay;
    this.factor = factor;
  }

  success() {
    return (this.current = this.noDelay);
  }

  fail() {
    return (this.current = Math.min(
      Math.max(this.minDelay, this.current * this.factor),
      this.maxDelay
    ));
  }
}

module.exports = Backoff;
