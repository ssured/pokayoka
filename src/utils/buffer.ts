// https://stackoverflow.com/a/53307879
export function toHex(buffer: ArrayBuffer) {
  let s = '';
  const h = '0123456789abcdef';
  new Uint8Array(buffer).forEach(v => {
    s += h[v >> 4] + h[v & 15];
  });
  return s;
}
