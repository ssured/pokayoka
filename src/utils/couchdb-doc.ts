import {
  observable,
  computed,
  when,
  action,
  autorun,
  reaction,
  runInAction,
} from 'mobx';
import { computedFn } from 'mobx-utils';
import { task, TaskStatusAware } from 'mobx-task';
import { generateId as globalGenerateId } from './id';
import { serializable, serialize, identifier, deserialize } from 'serializr';
import { async } from 'q';

const settings = observable({
  serverUrl: 'http://localhost:5984',
  database: 'atest',
  urlForDocument(id: string) {
    return [this.serverUrl, this.database, id].join('/');
  },
});

const getJson = task(async (url: string) => {
  const data = await fetch(url);
  return data.json();
});

const putJson = task(async (url: string, body: any) => {
  const rawResponse = await fetch(url, {
    method: 'PUT',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  return await rawResponse.json();
});

const api = observable({
  load: computedFn((id: string) => {
    const job = task(async () => {
      const url = settings.urlForDocument(id);
      console.log(`fetching ${url}`);
      const data = await fetch(url);
      const json = await data.json();
      return json;
    });
    job();
    return job;
  }),
  persist: (serialized: any) => {
    return putJson(settings.urlForDocument(serialized._id), serialized);
  },
});

function rawGet(identifier: string) {
  const [type, id] = identifier.split('-');

  const Class = type === 'project' ? NProject : NUser;

  const doc = task(async () => {
    const data = api.load(identifier);
    await when(() => !data.pending);
    if (data.error) {
      console.error('got loading error', data.error);
      throw data.error;
    }
    let object: any;
    runInAction(() => {
      object = deserialize(Class, data.result);
      object._rev = data.result._rev;
    });
    return object;
  });
  doc();
  return doc;
}
export const getNUser = computedFn(rawGet) as (
  id: string
) => TaskStatusAware<NUser, [string]>;

export class Document {
  static type = 'document';
  static generateId() {
    return `${this.type}-${globalGenerateId()}`;
  }

  @serializable(identifier())
  readonly _id?: string;

  @observable
  _rev?: string;

  constructor() {
    reaction(
      () => this.serialized,
      serialized => {
        // console.log('serialized: ' + JSON.stringify(this.serialized));
        api
          .persist({ _rev: this._rev, ...serialized })
          .then(action(({ ok, rev }) => ok && (this._rev = rev)));
      },
      { fireImmediately: false }
    );
  }

  @computed
  get remote() {
    return api.load(this._id);
  }

  @computed
  get serialized() {
    return serialize(this);
  }
}

export class NUser extends Document {
  static type = 'nuser';

  @serializable
  @observable
  name = '';

  @action setName(name: string) {
    this.name = name;
  }
}

export class NProject extends Document {
  static type = 'nproject';

  @observable
  name = '';
}
