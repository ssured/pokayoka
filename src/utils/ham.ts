// https://github.com/amark/gun/blob/1fe9daafdeeb55dc41966595c4e848a928a0fa02/src/HAM.js

type timestamp = string;

import { JsonEntry } from './json';
import Lexical from 'fast-json-stable-stringify';

type hamResult =
  | { resolution: 'defer' | 'historical' | 'equal' }
  | { resolution: 'merge'; incoming: boolean; current: boolean };

/* Based on the Hypothetical Amnesia Machine thought experiment */
export function ham(
  machineState: timestamp,
  incomingState: timestamp,
  currentState: timestamp,
  incomingValue: JsonEntry | undefined,
  currentValue: JsonEntry | undefined
): hamResult {
  if (machineState < incomingState) {
    return { resolution: 'defer' }; // the incoming value is outside the boundary of the machine's state, it must be reprocessed in another state.
  }
  if (incomingState < currentState) {
    return { resolution: 'historical' }; // the incoming value is within the boundary of the machine's state, but not within the range.
  }
  if (currentState < incomingState) {
    return { resolution: 'merge', incoming: true, current: false }; // the incoming value is within both the boundary and the range of the machine's state.
  }
  if (incomingState === currentState) {
    const incomingValueString = Lexical(incomingValue) || '';
    const currentValueString = Lexical(currentValue) || '';
    if (incomingValueString === currentValueString) {
      // Note: while these are practically the same, the deltas could be technically different
      return { resolution: 'equal' };
    }
    /*
			The following is a naive implementation, but will always work.
			Never change it unless you have specific needs that absolutely require it.
			If changed, your data will diverge unless you guarantee every peer's algorithm has also been changed to be the same.
			As a result, it is highly discouraged to modify despite the fact that it is naive,
			because convergence (data integrity) is generally more important.
			Any difference in this algorithm must be given a new and different name.
		*/
    if (incomingValueString < currentValueString) {
      // Lexical only works on simple value types!
      return { resolution: 'merge', incoming: false, current: true };
    }
    if (currentValueString < incomingValueString) {
      // Lexical only works on simple value types!
      return { resolution: 'merge', incoming: true, current: false };
    }
  }
  throw new Error(
    `Invalid CRDT Data ${incomingValue} to ${currentValue} at ${incomingState} to ${currentState}!`
  );
}
