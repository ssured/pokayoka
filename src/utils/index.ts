import { applyOperation } from 'fast-json-patch';

import isEqualWith from 'lodash.isequalwith';
export const isEqual = isEqualWith;

export const objectFromPatchStream = (): {
  object: () => any;
  updateObject: (msg: any) => void;
} => {
  let object: any = null;
  return {
    object: () => object,
    updateObject: (message: any) => {
      if (message.snapshot) {
        object = message.snapshot;
      } else if (message.patch) {
        applyOperation(object, message.patch, false, true);
      }
    },
  };
};

export const ensureNever = <T = any>(arg: never, shouldThrow = true): T => {
  if (shouldThrow) {
    throw new Error('Typescript never error, should never be reached');
  }
  return (undefined as unknown) as T;
};
