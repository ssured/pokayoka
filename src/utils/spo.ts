import dlv from 'dlv';

export type primitive = boolean | string | number | null | undefined;
export type subj = string[];
export type pred = string;
export type objt = primitive | subj;
export type state = string;
export type Tuple = [subj, pred, objt, state];
export type SPOShape = { [K in string]: primitive | SPOShape }; // links are hydrated
export type RawSPOShape = { [K in string]: objt | RawSPOShape }; // links are expressed as subj

const objToSubj = new WeakMap<RawSPOShape, subj>();

function subjsAreEqual(a: subj, b: subj) {
  if (a.length !== b.length) return false;
  return a.reduce((all, a, i) => all && a === b[i], false);
}

export function newRoot(): SPOShape {
  const newObject: SPOShape = {};
  objToSubj.set(newObject, []);
  return newObject;
}

function newObject(root: SPOShape, subj: subj): SPOShape {
  const parent = get(root, subj.slice(0, -1));
  parent[subj.slice(-1)[0]] = {};

  const newObject = get(root, subj); // trigger possible getters
  objToSubj.set(newObject, subj);

  return newObject;
}

export function getSubj(shape: SPOShape) {
  return objToSubj.get(shape);
}

export function get<P>(
  root: SPOShape,
  subj: subj,
  pred?: P
): P extends pred ? objt | SPOShape : SPOShape {
  let foundSubj = dlv<unknown>(root, subj);

  // if we encounter links, follow them with cycle checking
  const seen = new Set<unknown>([foundSubj]);
  while (isLink(foundSubj)) {
    if (subjsAreEqual(subj, foundSubj)) {
      // if you point to yourself you become empty
      return pred ? null : (newObject(root, subj) as any);
    }
    foundSubj = dlv<unknown>(root, foundSubj);
    if (seen.has(foundSubj)) {
      throw new Error('cycle in graph subjects');
    }
    seen.add(foundSubj);
  }

  if (typeof pred === 'string') {
    if (isSPOShape(foundSubj)) {
      const objt = foundSubj[pred] || null;
      if (isLink(objt)) return get(root, objt);
      return objt as any;
    }
    return null as any;
  }

  if (isSPOShape(foundSubj)) {
    return foundSubj as any;
  }

  return newObject(root, subj) as any;
}

export function set(
  root: SPOShape,
  subj: subj,
  pred: pred,
  objt: objt
): boolean {
  try {
    const parent = get(root, subj);

    if (objt == null) {
      delete parent[pred];
    }
    if (isLink(objt)) {
      parent[pred] = get(root, objt);
    } else {
      parent[pred] = objt;
    }
    return true;
  } catch (e) {
    return false;
  }
}

export function* spoInObject(
  subj: subj,
  obj: RawSPOShape,
  state: state
): Iterable<Tuple> {
  if (objToSubj.has(obj)) {
    throw new Error('this did not occur before, check implementation below');
    // console.log("frompaths", subj, obj, paths.get(obj));
    // if (subj.length === 0)
    //   throw new Error("impossible state, root references a known object");
    // yield {
    //   subj: subj.slice(0, -1),
    //   pred: subj.slice(-1)[0],
    //   objt: { "#": charwise.encode(paths.get(obj)) }
    // };
  } else {
    objToSubj.set(obj, subj);
    for (const [key, value] of Object.entries(obj)) {
      if (isObjt(value)) {
        yield [subj, key, value, state];
      } else if (isObject(value)) {
        if (Array.isArray(value)) {
          throw new Error(
            `arrays are not supported ${JSON.stringify([subj, key, value])}`
          );
        }
        if (objToSubj.has(value)) {
          yield [subj, key, objToSubj.get(value)!, state];
        } else {
          yield* spoInObject(subj.concat(key), value, state);
          // yield [subj, key, subj.concat(key)];
        }
      }
    }
  }
}

export function isObject(x: unknown): x is object {
  return (typeof x === 'object' && x !== null) || typeof x === 'function';
}

export function isObjt(v: unknown): v is objt {
  switch (typeof v) {
    case 'string':
    case 'boolean':
    case 'number':
      return true;
    case 'object':
      return v == null || isLink(v);
  }
  return false;
}

export function isLink(x: unknown): x is subj {
  // TODO in production this check might be too expensive
  if (!Array.isArray(x) || x.length === 0) return false;
  for (let i = 0; i < x.length; i += 1) {
    if (typeof x[i] !== 'string') return false;
  }
  return true;
}

export function isSPOShape(o: unknown, deep: boolean = false): o is SPOShape {
  return (
    isObject(o) &&
    !Array.isArray(o) &&
    (!deep ||
      Object.values(o).reduce(
        (all, value) => all && (isObjt(value) || isSPOShape(value)),
        true
      ))
  );
}
