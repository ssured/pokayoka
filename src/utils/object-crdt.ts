import { mergeWith } from 'lodash';
import { isObject } from './spo';

type objt = null | string | number | boolean | string[];

type MergeableShape<T extends object> = {
  [P in keyof T]?: T[P] extends object ? MergeableShape<T[P]> : objt
};
export interface IMergeable extends MergeableShape<object> {}

export type ToMergeableObject<T extends IMergeable> = {
  [K in keyof T]: Record<
    string,
    T[K] extends IMergeable ? ToMergeableObject<T[K]> : T[K]
  >
};

export type FromMergeableObject<
  T extends ToMergeableObject<IMergeable>
> = T extends ToMergeableObject<infer U> ? U : never;

type Value<T> = T[keyof T];

const isObjt = (v: unknown): v is objt => {
  if (v == null) return v === null;
  return typeof v !== 'object' || Array.isArray(v);
};

export function merge<T extends IMergeable>(
  object: ToMergeableObject<T>,
  ...sources: ToMergeableObject<T>[]
): ToMergeableObject<T> {
  return mergeWith(object, ...sources, (
    objValue: Value<ToMergeableObject<T>>,
    srcValue: Value<ToMergeableObject<T>>,
    key: string,
    object: ToMergeableObject<T>,
    source: ToMergeableObject<T>
    //   stack: unknown
  ) => {
    if (
      key in object &&
      key in source &&
      isObjt(objValue) &&
      isObjt(srcValue)
    ) {
      // resolve the conflict
      return (JSON.stringify(objValue) || '') > (JSON.stringify(srcValue) || '')
        ? objValue
        : srcValue;
    }
  });
}

const descending = (a: string, b: string) => (a < b ? 1 : -1);

/**
 *
 * @param state Which state should the value be wrapped with
 * @param value The value to wrap
 * @returns Object of the value at the state
 */
export function asMergeableObject<
  S extends string,
  T extends MergeableShape<any> | objt
>(
  state: S,
  value: T
): T extends IMergeable ? ToMergeableObject<T> : { [K in S]: T } {
  return {
    [state]: isObjt(value)
      ? value
      : Object.entries(value as MergeableShape<any>).reduce(
          (map, [key, value]) => {
            map[key] = value ? asMergeableObject(state, value as any) : null;
            return map;
          },
          {} as any
        ),
  } as any;
}

export function valueAt<T extends IMergeable>(
  state: string,
  object: ToMergeableObject<T>
): T {
  return Object.entries(object as Record<
    string,
    Record<string, unknown>
  >).reduce(
    (object, [key, stateMap]) => {
      const states = Object.keys(stateMap).sort(descending);
      const currentState = states.find(k => k <= state) || '';
      const currentValue = stateMap[currentState];

      if (currentValue === undefined) {
        object[key] = null;
      } else if (isObjt(currentValue)) {
        object[key] = currentValue;
      } else {
        object[key] = valueAt(state, currentValue as any);
      }

      return object;
    },
    {} as any
  );
}
