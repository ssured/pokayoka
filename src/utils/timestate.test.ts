import charwise from 'charwise';
import {
  action,
  autorun,
  configure,
  observable,
  runInAction,
  IObservableValue,
  ObservableMap,
  toJS,
  reaction,
  when,
} from 'mobx';
import { isObject } from 'util';
import { generateId } from '../../server/utils/snag-id';
import {
  checkDefinitionOf,
  many,
  serializeMany,
  UniversalObject,
  MergableSerialized,
  mergeSerialized,
  Serialized,
  setPathOf,
  serializeOne,
} from './object-live-crdt';
import {
  createEmittingRoot,
  RootEventMsg,
  createRoot,
} from './observable-root';
import { isLink, subj } from './spo';
import SubscribableEvent, { SubscriptionToken } from 'subscribableevent';
import { ensureNever } from './index';
import { valueAt, IMergeable } from './object-crdt';

configure({ enforceActions: 'always', disableErrorBoundaries: true });

// state is global for the system
let state!: IObservableValue<number>;
const getState = () => String(state.get());

let timeState!: ObservableMap<
  string,
  Record<string, MergableSerialized<AnyInstance>>
>;

runInAction(() => {
  timeState = observable.map<
    string,
    Record<string, MergableSerialized<AnyInstance>>
  >([
    [
      path2key(['existing@example.com']),
      {
        '0': {
          name: {
            '0': 'Existing',
          },
          projects: {
            '0': {},
          },
        },
      } as any,
    ],
  ]);
});

function getTimeStateThread(identifier: string) {
  if (!timeState.has(identifier)) {
    runInAction(() => timeState.set(identifier, { '0': {} } as any));
  }
  return timeState.get(identifier)!;
}
function isTopLevel(key: string) {
  return key2path(key).length === 1;
}

abstract class Base extends UniversalObject {
  constructor(readonly identifier: string) {
    super();

    // only sync on top level objects
    // deeper objects will be synced from top
    if (!isTopLevel(identifier)) return;

    // TODO remove path logic and use identifier for all paths
    setPathOf(this, key2path(identifier));
  }
}

@checkDefinitionOf<Project>()
class Project extends Base {
  static '@type' = 'Project';

  static serialize({ name }: Project) {
    return { name };
  }

  @observable
  name = '';

  @action
  setName(name: string) {
    this.name = name;
  }
}

@checkDefinitionOf<User>()
class User extends Base {
  static '@type' = 'User';

  static constructors = {
    projects: many(Project),
    defaultProject: Project,
  };

  static serialize(user: User) {
    const { name } = user;
    return {
      name,
      ...serializeMany(user, 'projects'),
      ...serializeOne(user, 'defaultProject'),
    };
  }

  @observable
  name = '';

  @action
  setName(name: string) {
    this.name = name;
  }

  @observable
  defaultProject?: Project;

  projects = observable.map<string, Project>();

  @action
  addProject(id: string, name: string) {
    const key = path2key([this.identifier, 'projects', id]);
    const project = new Project(key);
    project.setName(name);
    this.projects.set(project.identifier, project);
    return project;
  }

  @action
  selectProject(project?: Project) {
    this.defaultProject = project;
  }
}

// Helpers for the shape of internal state
type AnyClass = typeof User | typeof Project;
type AnyInstance = InstanceType<AnyClass>;
const rootShape = many(User);
const lookupClassFromPath = (path: subj): AnyClass => {
  if (path.length === 0) throw new Error('root has no constructor');
  let ctor = rootShape as any;

  for (const part of path) {
    if (!isObject(ctor.constructors)) {
      throw new Error(
        `could not find constructors at path ${JSON.stringify(path)}`
      );
    }
    ctor = ctor.constructors[part];
    if (!ctor || typeof ctor['@type'] !== 'string') {
      throw new Error(
        `no or wrong constructor at path ${JSON.stringify(path)}`
      );
    }
  }

  return ctor;
};
function key2path(key: string): subj {
  let path!: subj | unknown;
  try {
    path = charwise.decode(key) as unknown;
  } catch (e) {
    throw new Error(`Key is properly encoded: ${key}`);
  }
  if (!isLink(path)) {
    throw new Error(`key is not a valid path ${JSON.stringify(key)}`);
  }
  return path;
}
function path2key(path: subj): string {
  return charwise.encode(path);
}

const timeStateHandler = () => {
  const disposersMap: Record<string, (() => void)[]> = {};
  const instancesMap = observable.map<string, AnyInstance>();
  return async (msg: RootEventMsg<AnyInstance>) => {
    const { key } = msg;
    if (!isTopLevel(key)) return;

    switch (msg.type) {
      case 'observed': {
        const store = getTimeStateThread(key);

        let mutexLocked = false;

        disposersMap[key] = [];
        disposersMap[key].push(
          reaction(
            () => valueAt(getState(), store) as Partial<Serialized<any>>,
            current => {
              const instance = instancesMap.get(key);
              if (current == null || instance == null || mutexLocked) {
                return;
              }
              const Class = instance.constructor as any;

              try {
                mutexLocked = true;
                Class.merge(instance, current);
              } finally {
                mutexLocked = false;
              }
            },
            { fireImmediately: true }
          )
        );

        disposersMap[key].push(
          reaction(
            () => {
              const instance = instancesMap.get(key);
              if (instance == null) return;
              const Class = instance.constructor as any;
              return Class.serialize(instance);
            },
            serialized => {
              if (serialized == null || mutexLocked) return;
              try {
                mutexLocked = true;
                mergeSerialized(getState(), store, serialized);
              } finally {
                mutexLocked = false;
              }
            },
            { fireImmediately: false }
          )
        );

        break;
      }

      case 'unobserved':
        disposersMap[key].forEach(disposer => disposer());
        break;
      case 'update':
        runInAction(() => instancesMap.set(key, msg.value));
        break;
      default:
        ensureNever(msg);
    }
  };
};

describe('one class', () => {
  type Message = RootEventMsg<AnyInstance>;
  let root!: Record<string, AnyInstance>;
  let subscribe!: (callback: (msg: Message) => void) => SubscriptionToken;

  beforeEach(() => {
    state = observable.box(1);
    const rootBag = createEmittingRoot<AnyInstance>({
      onObserved: (key, set) =>
        runInAction(() => {
          const Class = lookupClassFromPath(key2path(key));
          const instance = new Class(key);
          set(instance);
        }),
    });
    root = rootBag.root;
    subscribe = rootBag.subscribe;
  });

  test('it allows external updates', async () => {
    subscribe(timeStateHandler());

    const username = 'test@example.com';
    const key = path2key([username]);

    const keepAlive = when(() => root[key].name === 'done');

    const user = root[key] as User;
    expect(user.name).toBe('');

    runInAction(() => state.set(2));
    user.setName('Two');
    expect(toJS(timeState)).toMatchSnapshot();
    // console.log(toJS(globalStore));

    runInAction(() => state.set(3));
    user.setName('Three');
    expect(toJS(timeState)).toMatchSnapshot();

    mergeSerialized('4', getTimeStateThread(key), { name: 'Four' });

    expect(user.name).toBe('Three');
    runInAction(() => state.set(4));
    expect(user.name).toBe('Four');

    setTimeout(() => root[key]!.setName('done'), 10);

    // expect(1).toBe(2)

    await keepAlive;
    console.log('done');
  });

  test('it supports one to many', async () => {
    subscribe(timeStateHandler());

    const username = 'existing@example.com';
    const key = path2key([username]);

    const keepAlive = when(() => root[key].name === 'done');

    const user = root[key] as User;
    expect(user.name).toBe('Existing');

    runInAction(() => state.set(2));
    user.addProject('p1', 'Two');
    expect(toJS(timeState)).toMatchSnapshot();

    runInAction(() => state.set(3));
    user.projects
      .values()
      .next()
      .value.setName('Three');
    expect(toJS(timeState)).toMatchSnapshot();

    setTimeout(() => root[key]!.setName('done'), 10);

    // expect(1).toBe(2)

    await keepAlive;
    console.log('done');
  });

  test.only('it supports zero or one', async () => {
    subscribe(timeStateHandler());

    const username = 'existing@example.com';
    const key = path2key([username]);

    const keepAlive = when(() => root[key].name === 'done');

    const user = root[key] as User;
    expect(user.name).toBe('Existing');

    runInAction(() => state.set(2));
    const project = user.addProject('p1', 'Two');

    expect(toJS(timeState)).toMatchSnapshot();

    runInAction(() => state.set(3));
    user.projects
      .values()
      .next()
      .value.setName('Three');
    user.selectProject(project);
    expect(toJS(timeState)).toMatchSnapshot();

    setTimeout(() => root[key]!.setName('done'), 10);

    // expect(1).toBe(2)

    await keepAlive;
    console.log('done');
  });
});
