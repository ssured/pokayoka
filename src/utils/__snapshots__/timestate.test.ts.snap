// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one class it supports zero or one 1`] = `
Object {
  "KJexisting@example.com!": Object {
    "0": Object {
      "defaultProject": Object {
        "2": null,
      },
      "name": Object {
        "0": "Existing",
      },
      "projects": Object {
        "0": Object {
          "KJKJexisting@example.com??\\"Jprojects\\"Jp1!": Object {
            "2": Object {
              "@type": Object {
                "2": "Project",
              },
              "identifier": Object {
                "2": "KJKJexisting@example.com??\\"Jprojects\\"Jp1!",
              },
              "name": Object {
                "2": "Two",
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`one class it supports zero or one 2`] = `
Object {
  "KJexisting@example.com!": Object {
    "0": Object {
      "defaultProject": Object {
        "2": null,
        "3": Array [
          "existing@example.com",
          "projects",
          "KJKJexisting@example.com??\\"Jprojects\\"Jp1!",
        ],
      },
      "name": Object {
        "0": "Existing",
      },
      "projects": Object {
        "0": Object {
          "KJKJexisting@example.com??\\"Jprojects\\"Jp1!": Object {
            "2": Object {
              "@type": Object {
                "2": "Project",
              },
              "identifier": Object {
                "2": "KJKJexisting@example.com??\\"Jprojects\\"Jp1!",
              },
              "name": Object {
                "2": "Two",
                "3": "Three",
              },
            },
          },
        },
      },
    },
  },
}
`;
